frameworks/Java/play2-java/play2-java-jooq-hikaricp/app/controllers/Application.java

```diff
@@ -0,0 +1,121 @@
+        this.db = db;
+        this.dbEc = dbEc;
+    }
+
+    public CompletionStage<Result> db() {
+        return getRandomWorlds(1).thenApply(worlds -> ok(worlds.get(0).formatJSON(JSON_FORMAT)).as(JSON));
+    }
+
+    public CompletionStage<Result> queries(final String queries) {
+        return getRandomWorlds(queryCount(queries)).thenApply(worlds -> ok(worlds.formatJSON(JSON_FORMAT)).as(JSON));
+    }
+
+    public CompletionStage<Result> fortunes() {
+        return CompletableFuture.supplyAsync(() -> {
+            final List<FortuneRecord> fortunes = this.db.withConnection(connection -> {
+                return DSL.using(connection, DIALECT).selectFrom(FORTUNE).fetch();
+            });
+            fortunes.add(new FortuneRecord(UInteger.valueOf(0), "Additional fortune added at request time."));
+            Collections.sort(fortunes, (f1, f2) -> f1.getMessage().compareTo(f2.getMessage()));
+
+            return ok(views.html.fortunes.render(fortunes));
+        }, dbEc);
+    }
+
+    public CompletionStage<Result> update(final String queries) {
+        return getRandomWorlds(queryCount(queries)).thenApplyAsync(worlds -> {
+            final Random random = ThreadLocalRandom.current();
+            for (final WorldRecord world : worlds) {
+                world.setRandomnumber((random.nextInt(10000) + 1));
+            }
+
+            final int batchSize = 25;
+            final int batches = ((worlds.size() / batchSize) + 1);
+            this.db.withConnection(connection -> {
+                final DSLContext sql = DSL.using(connection, DIALECT);
+                for ( int i = 0 ; i < batches ; ++i ) {
+                    sql.batchUpdate(worlds.subList(i * batchSize, Math.min((i + 1) * batchSize, worlds.size()))).execute();
+                }
+                return null;
+            });
+
+            return ok(worlds.formatJSON(JSON_FORMAT)).as(JSON);
+        }, dbEc);
+    }
+
+    private int queryCount(final String queryCountString) {
+        int queryCount;
+        try {
+            queryCount = Integer.parseInt(queryCountString, 10);
+        } catch (final NumberFormatException e) {
+            queryCount = 1;
+        }
+        if (queryCount < 1) {
+            queryCount = 1;
+        } else if (queryCount > 500) {
+            queryCount = 500;
+        }
+
+        return queryCount;
+    }
+
+    private CompletionStage<org.jooq.Result<WorldRecord>> getRandomWorlds(final int n) {
+        return CompletableFuture.supplyAsync(() -> {
+            final Random random = ThreadLocalRandom.current();
+            org.jooq.Result<WorldRecord> worlds = null;
+            for (int i = 0; i < n; ++i) {
+                long randomId = random.nextInt(10000) + 1;
+                final org.jooq.Result<WorldRecord> world = this.db.withConnection(connection -> {
+                    return DSL.using(connection, DIALECT).selectFrom(WORLD).where(WORLD.ID.eq(UInteger.valueOf(randomId))).fetch();
```

Comment:
This is again something specific to benchmarks, and not the real world, but if you run thousands of individual selects like this, you could save some time by generating the SQL string and bind variables only once, storing that in some local variable, and then executing it several times.

JPA implementations can easily cache their trivial CRUD queries, because they are trivial ("fetch by ID"). It would be much more interesting to benchmark more complex dynamic SQL (e.g. jOOQ vs. Criteria API)...

Reply:
Test requirement says there needs to be a database round trip for each fetch.
Do you think there is a major impact when creating the jOOQ query each time? Can that one be cached on client side?

Reply:
Again, there isn't a major impact in a *real world* application, but there can be a significant impact in a *benchmark*.

Nevertheless, in this particular comment, I didn't advocate for preventing the database round trip per fetch, but the SQL generation per fetch. You can easily do:

```java
ResultQuery<?> q1 = ...
String sql = q1.getSQL();
List<Object> binds = q1.getBindValues();

ResultQuery q2 = DSL.using(...).resultQuery(sql, binds.toArray());

// and now loop and reuse q2
```

Again, that's not something that is advisable for every day business logic, where the improvement should be non-negligible, but this is a *benchmark*. It measures *mainly* this overhead many many times.